### ISSUE TITLE
Message Decoder Conflict

### ISSUE THREAD

**Issue by martintreurnicht** [ethereum/ethereumj#929]
I'm playing around with whisper at the moment and i noticed it gets a `NullPointerException` as soon as whisper is activated, it seems that the `MessageCodec` is where the problem is, it seems that the `frameType`, which is `28` in this case, is both valid for `eth` and `shh`, when resolving the message code from Whisper, it returns `0` which is indicates that it's a `STATUS` message, but it also returns in range for `eth` as `12` (`0x0c`), which indicates as in range, but i can't find a code for it

For now i'm guessing the easiest fix would be to move the whisper decoder above the eth decoder, but i'm not sure if this would cause other errors.

```bash
java.lang.NullPointerException: null
        at org.ethereum.net.eth.handler.Eth63.processNodeData(Eth63.java:193)
        at org.ethereum.net.eth.handler.Eth63.channelRead0(Eth63.java:90)
        at org.ethereum.net.eth.handler.Eth63.channelRead0(Eth63.java:55)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:108)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at org.ethereum.net.rlpx.NettyByteToMessageCodec.channelRead(NettyByteToMessageCodec.java:73)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at org.ethereum.net.server.WireTrafficStats$TrafficStatHandler.channelRead(WireTrafficStats.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by martintreurnicht** [comment]
https://github.com/ethereum/ethereumj/pull/931, although a bit of a hack, does resolve this issue

**Comment by mlahia** [comment]
Looking at the code it seems the failure is related to the fact that `requestedNodes` is not checked to be `null` before calling `.contains` on it. My naive move would be to add a non null check. At least to avoid the exception in this case.

**Comment by martintreurnicht** [comment]
@mlahia I commented on your commit. Yes there is a Nullpointer check that can be added, but that doesn't really solve the underlying issue. The core issue is that the message is being identified as the wrong type, thus causing it to go down a channel that it's not suppose to causing all types of other issues along the way. Have a look at my pull request to see what the root issue is

**Comment by mlahia** [comment]
Yeah, I only tried to prevent the null pointer error. I don't yet know the whole process to be able to spot the real cause. Though it could be helpfull if I can make a failing unit-test that reproduce the problem.

Note also that my fix, which appears to be not correct (as per your comment) did not make any test fail, which bothers me a little.

@martintreurnicht do you think we can reproduce the issue in a unit-test easily?

Thanks for your comments, I'll try to dig in more deeper.

**Comment by mlahia** [comment]
Looking at the code on `MessageCodeResolver.java`, it seems that P2P offset is not set on the `init` method. So I suspect that this might be the root cause and that we should add something like this at the beggining of the for loop:

```java
            if (capability.getName().equals(Capability.P2P)) {
                setOffset(Capability.P2P, offset);
                offset += EthMessageCodes.values().length;
            }
```
@martintreurnicht what do you think ?

**Comment by mlahia** [comment]
I think I found out the root issue, it seems that `inRange` method in `EthMessageCode` class is not propertly checking that a `code` is part of the supported list of codes.

@martintreurnicht can you please check my latest update to the code fix?

**Comment by martintreurnicht** [comment]
@mlahia That is correct. I've tracked down the root issue, tested it and updated the code to reflect that change rather. This also removes the need for the FIXME that use to be in there for manually adjusting the offset for swarm

**Comment by martintreurnicht** [comment]
@mlahia I will see if i can add a test to prove this issue. I think the reason why it's not showing up on any of the tests is because they are all running in isolation, meaning they don't switch on more than one protocol per test

**Comment by martintreurnicht** [comment]
Ok, so the reason why the tests aren't failing is because no one ever upgraded the tests for ETH63, also all the tests do a offset test, but it doesn't verify that offsets with the `inRange` check. For instance if you look at `AdaptiveMessageIdsTest.test6()` that i just added and ran it against the old code you'll see it fails, specifically this piece of code:

```java
        byte shhCode = messageCodesResolver.withShhOffset(ShhMessageCodes.STATUS.asByte());
        byte ethCode = messageCodesResolver.resolveEth(shhCode);
        assertFalse(EthMessageCodes.inRange(ethCode, EthVersion.V63));
```

How can a shh code resolve to a eth code?

**Comment by martintreurnicht** [comment]
@Nashatyrev Related to pull request #931

**Comment by mlahia** [comment]
@martintreurnicht I've run your new test with the fix I did, and it is passing. Here is the test:

```java
    @Test
    public void ShhCodeShouldNotBeResolvedAsEthOne() {

        List<Capability> capabilities = Arrays.asList(
                new Capability(Capability.ETH, EthVersion.V63.getCode()),
                new Capability(Capability.SHH, ShhHandler.VERSION));

        messageCodesResolver.init(capabilities);

        byte shhCode = messageCodesResolver.withShhOffset(ShhMessageCodes.STATUS.asByte());
        byte ethCode = messageCodesResolver.resolveEth(shhCode);
        assertFalse(EthMessageCodes.inRange(ethCode, EthVersion.V63));
    }
```
Thanks for the test, this helped me validate that my fix in https://github.com/ethereum/ethereumj/pull/935 is working.

**Comment by mkalinin** [comment]
Resolved in [1.7.0](https://github.com/ethereum/ethereumj/releases/tag/1.7.0)
